File: 63.不同路径-ii.py
#
# @lc app=leetcode.cn id=63 lang=python
#
# [63] 不同路径 II
#
# 解法1(T82% S90%): 上一道题的约束变种，在传播过程中只要遇到有石头的就continue跳过即可，比较复杂的是初始化阶段，要从右下角向上和向左初始化两条边，如果遇到石头就break不初始化上面和左边一段，此外还要判断左上角和右下角是不是有石头

# @lc code=start
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        m, n = len(obstacleGrid), len(obstacleGrid[0])

        if obstacleGrid[0][0] or obstacleGrid[m-1][n-1]:
            return 0

        dp = [[0]*n for _ in range(m)]
        for i in range(m-1, -1, -1):
            if obstacleGrid[i][n-1]:
                break
            dp[i][n-1] = 1
        for j in range(n-1, -1, -1):
            if obstacleGrid[m-1][j]:
                break
            dp[m-1][j] = 1

        for i in range(m-2, -1, -1):
            for j in range(n-2, -1, -1):
                if obstacleGrid[i][j]:
                    continue
                dp[i][j] = dp[i+1][j] + dp[i][j+1]
        
        return dp[0][0]
# @lc code=end

File: javascript/63.不同路径-ii.js
/*
 * @lc app=leetcode.cn id=63 lang=javascript
 *
 * [63] 不同路径 II
 */

// @lc code=start
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
    var m=obstacleGrid.length, n=obstacleGrid[0].length;
    if (obstacleGrid[0][0]==1 || obstacleGrid[m-1][n-1]==1) { return 0; }

    var dp = Array.from(new Array(m).fill(0), () => new Array(n).fill(0));
    // 这里不需要反过来考虑，如果第一列中间有一个石头，则石头下面的永远也走不到终点，石头上面的还有一次机会
    for (var i=0; i<m; i++) {
        if (obstacleGrid[i][0] != 1) { dp[i][0] = 1; }
        else { break; }
    }
    for (var j=0; j<n; j++) {
        if (obstacleGrid[0][j] != 1) { dp[0][j] = 1; }
        else { break; }
    }
    for (var i=1; i<m; i++) {
        for (var j=1; j<n; j++) {
            if (obstacleGrid[i][j] != 1) {
                dp[i][j] = dp[i-1][j] + dp[i][j-1];
            }
        }
    }
    return dp[m-1][n-1];
};
// @lc code=end

